boto/endpoints.py

```diff
@@ -0,0 +1,200 @@
+
+
+class BotoEndpointResolver(EndpointResolver):
+    """Endpoint resolver which handles boto2 compatibility concerns."""
+
+    SERVICE_RENAMES = {
+        # The botocore resolver is based on endpoint prefix.
+        # These don't always sync up to the name that boto2 uses.
+        # A mapping can be provided that handles the mapping between
+        # "service names" and endpoint prefixes.
+        'awslambda': 'lambda',
+        'cloudwatch': 'monitoring',
+        'ses': 'email',
+        'ec2containerservice': 'ecs',
+        'configservice': 'config',
+    }
+
+    def __init__(self, endpoint_data, service_rename_map=None):
+        """
+        :type endpoint_data: dict
+        :param endpoint_data: Regions and endpoints data in the same format
+            as is used by botocore / boto3.
+
+        :type legacy_endpoint_data: dict
+        :param legacy_endpoint_data: Regions and endpoints data in the legacy
+            format. This data takes precedence over any data found in
+            `endpoint_data`.
+
+        :type service_rename_map: dict
+        :param service_rename_map: A mapping of boto2 service name to
+            endpoint prefix.
+        """
+        super(BotoEndpointResolver, self).__init__(endpoint_data)
+        if service_rename_map is None:
+            service_rename_map = self.SERVICE_RENAMES
+        # Mapping of boto2 service name to endpoint prefix
+        self._endpoint_prefix_map = service_rename_map
+        # Mapping of endpoint prefix to boto2 service name
+        self._service_name_map = dict(
+            (v, k) for k, v in service_rename_map.items())
+
+    def get_available_endpoints(self, service_name, partition_name='aws',
+                                allow_non_regional=False):
+        endpoint_prefix = self._endpoint_prefix(service_name)
+        return super(BotoEndpointResolver, self).get_available_endpoints(
+            endpoint_prefix, partition_name, allow_non_regional)
+
+    def get_all_available_regions(self, service_name):
+        """Retrieve every region across partitions for a service."""
+        regions = set()
+        endpoint_prefix = self._endpoint_prefix(service_name)
+
+        # Get every region for every partition in the new endpoint format
+        for partition_name in self.get_available_partitions():
+            if self._is_global_service(service_name, partition_name):
+                # Global services are available in every region in the
+                # partition in which they are considered global.
+                partition = self._get_partition_data(partition_name)
+                regions.update(partition['regions'].keys())
+            else:
+                regions.update(
+                    super(BotoEndpointResolver, self).get_available_endpoints(
+                        endpoint_prefix, partition_name
+                    )
+                )
+
+        return list(regions)
+
+    def construct_endpoint(self, service_name, region_name=None):
```

Comment:
Ah, I was about to ask about this.  In that case, why is this a subclass of `BotoEndpointResolver` if we're adding additional methods that (presumably) are going to be the preferred methods to use with this class (i.e `resolve_hostname` instead of `construct_endpoint`).  Seems simpler to just use delegation and only expose the methods that are needed.